### PR TITLE
Adds support for retrieving credentials in the Cloud Shell environment.

### PR DESCRIPTION
**Pull request by wilford**

### CODE REVIEW

**Comment by anthmgoogle** [comment]
LGTM.

File: google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    try {    
+      PrintWriter out =
+        new PrintWriter(socket.getOutputStream(), true);
+      out.println(GET_AUTH_TOKEN_REQUEST);
+
+      BufferedReader input =
+          new BufferedReader(new InputStreamReader(socket.getInputStream()));
+      String ignoredSizeLine = input.readLine();
+      Collection<String> messageArray = GsonFactory.getDefaultInstance()
+	.createJsonParser(input)
```
Comment (ejona86) [inline google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java]: Could we please not use tabs for indentation?

File: google-api-client/pom.xml [hunk 2]
```diff
@@ -110,7 +111,6 @@
     <dependency>
       <groupId>com.google.http-client</groupId>
       <artifactId>google-http-client-gson</artifactId>
-      <scope>test</scope>
```
Comment (ejona86) [inline google-api-client/pom.xml]: This still adds a new dependency on gson. I think you may want to use the JsonFactory interface from google-http-java-client.

File: google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java (hunk 1 above)
Comment (wilford) [inline google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java]: Done.

File: google-api-client/pom.xml (hunk 2 above)
Comment (wilford) [inline google-api-client/pom.xml]: JsonFactory is an interface though, so which implementing class should I use?  It seems that any class I use adds a dependency?

**Comment by wilford** [comment]
Which implementing class of JsonFactory should I use?  It seems that all of them would add another dependancy?

File: google-api-client/pom.xml (hunk 2 above)
Comment (ejona86) [inline google-api-client/pom.xml]: A cursory look around seems like code just expects it to be passed in as a parameter. There does exist `Util.getDefaultJsonFactory()`, and it seems there is already a dependency on Jackson. However, the only place it is used is in `GoogleCredential`.

https://github.com/google/google-api-java-client/blob/57fe35766cbba0a0d5a9a296be81468d730a29f8/google-api-client/src/main/java/com/google/api/client/googleapis/compute/ComputeCredential.java#L70
https://github.com/google/google-api-java-client/blob/57fe35766cbba0a0d5a9a296be81468d730a29f8/google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/GooglePublicKeysManager.java#L93
https://github.com/google/google-api-java-client/blob/57fe35766cbba0a0d5a9a296be81468d730a29f8/google-api-client/src/main/java/com/google/api/client/googleapis/json/GoogleJsonResponseException.java#L92
https://github.com/google/google-api-java-client/blob/57fe35766cbba0a0d5a9a296be81468d730a29f8/google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/GoogleClientSecrets.java#L190
https://github.com/google/google-api-java-client/blob/57fe35766cbba0a0d5a9a296be81468d730a29f8/google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/GoogleAuthorizationCodeTokenRequest.java#L90
https://github.com/google/google-api-java-client/blob/57fe35766cbba0a0d5a9a296be81468d730a29f8/google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/GoogleRefreshTokenRequest.java#L84

File: google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    try {    
+      PrintWriter out =
+        new PrintWriter(socket.getOutputStream(), true);
+      out.println(GET_AUTH_TOKEN_REQUEST);
+
+      BufferedReader input =
+          new BufferedReader(new InputStreamReader(socket.getInputStream()));
+      String ignoredSizeLine = input.readLine();
+      Collection<String> messageArray = GsonFactory.getDefaultInstance()
+<<<<<<< HEAD
```
Comment (anthmgoogle) [inline google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java]: Incomplete merge?

File: google-api-client/pom.xml (hunk 2 above)
Comment (anthmgoogle) [inline google-api-client/pom.xml]: Yes, you should take the JsonFactory as a parameter to CloudShellCredential, and use the one that getApplicationDefault has had passed in. If you want to have a default value on CloudShellCredential for cases where people use it directly, use Util.getDefaultJsonFactory().

File: google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java (hunk 3 above)
Comment (wilford) [inline google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/CloudShellCredential.java]: Done.

File: google-api-client/pom.xml (hunk 2 above)
Comment (wilford) [inline google-api-client/pom.xml]: Done.

**Comment by wilford** [comment]
Fixed deps issues.

**Comment by ejona86** [comment]
This LGTM. @anthmgoogle still look good to you?

@wilford if you're ready and @anthmgoogle LGTMs (because I consider mine more of OWNERS approval), then I'll go ahead and merge it.

**Comment by wilford** [comment]
I'm ready whenever @anthmgoogle <https://github.com/anthmgoogle>  is.
Thanks!

**Comment by anthmgoogle** [comment]
LGTM. Thanks. We should be ready to merge now.

**Comment by ejona86** [comment]
@wilford, checkstyle fails. Please run `mvn verify` and fix the style failures.

**Comment by wilford** [comment]
Ah, my bad.  Fixed checkstyle errors.

**Comment by ejona86** [comment]
Cherry-picked as 3d866eb7

**Pull request closed without merging**